File: DeepSpace-Ruby/lib/Hangar.rb
# encoding:utf-8
require_relative "WeaponType"
require_relative "Weapon"
require_relative "ShieldBooster"
require_relative "HangarToUI"

module Deepspace
class Hangar
    attr_reader :maxElements, :weapons, :shieldBoosters

    def initialize (capacity)
        @maxElements    = capacity
        @weapons        = Array.new
        @shieldBoosters = Array.new
    end

    def self.newCopy (h)
        auxiliar = Hangar.new(h.maxElements)

        h.shieldBoosters.each { |escudos| auxiliar.addShieldBooster(escudos) }
        h.weapons.each { | armas | auxiliar.addWeapon(armas)}

        auxiliar
    end

    def shields
        @shieldBoosters
    end

    def spaceAvailable
        @shieldBoosters.size + @weapons.size < @maxElements
    end

    def addWeapon (w)
        if spaceAvailable
            @weapons.push(w)
            true
        else
            false
        end
    end

    def addShieldBooster (s)
        if spaceAvailable
            @shieldBoosters.push(s)
            true
        else
            false
        end
    end

    def removeShieldBooster (s)
        @shieldBoosters.delete_at(s)
    end

    def removeWeapon (w)
        @weapons.delete_at(w)
    end

    def getUIversion
        HangarToUI.new(self)
    end

    def to_s
        "\t-> Elementos máximos: #{@maxElements} \n\t-> Tamaño armas: #{@weapons.size} \n\t-> Tamaño escudos: #{@shieldBoosters.size}"
    end
end
end

File: DeepSpace-Java/src/deepspace/Hangar.java
package deepspace;

import java.util.ArrayList;

class Hangar implements Copyable <Hangar> {
    private int maxElements;
    private ArrayList<Weapon> weapons        = new ArrayList<>();
    private ArrayList<ShieldBooster> shields = new ArrayList<>();


    Hangar (int capacity) {
        maxElements = capacity;
    }

    Hangar (Hangar h) {
        maxElements = h.maxElements;
        weapons     = h.weapons;
        shields     = h.shields;
    }

    @Override
    public Hangar copy() {
        return new Hangar(this);
    }

//
// ──────────────────────────────────────────────────────────── MODIFICADORES ─────
//


    public boolean addWeapon (Weapon w) {
        if (SpaceAvailable()) {
            weapons.add(w);
            return true;
        }
        else {
            return false;
        }
    }

    public boolean addShieldBooster (ShieldBooster s) {
        if (SpaceAvailable()) {
            shields.add(s);
            return true;
        }
        else {
            return false;
        }
    }

//
// ────────────────────────────────────────────────────────────── CONSULTORES ─────
//

    public int getMaxElements () {
        return maxElements;
    }

    public ArrayList<ShieldBooster> getShieldBoosters () {
        return shields;
    }

    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }

    private boolean SpaceAvailable () {
        return shields.size() + weapons.size() < maxElements;
    }

//
// ─────────────────────────────────────────────────────────────────── REMOVE ─────
//

    public ShieldBooster removeShieldBooster (int s) {
        if (s < shields.size())
            return shields.remove(s);
        else
            return null;
    }

    public Weapon removeWeapon (int w) {
        if (w < weapons.size())
            return weapons.remove(w);
        else
            return null;
    }

    HangarToUI getUIversion () {
        return new HangarToUI(this);
    }

    public String toString() {
        return getUIversion.toString;;
    }
}
